apps/remote_control/lib/lexical/remote_control/build/parse_error.ex

```diff
@@ -0,0 +1,176 @@
+defmodule Lexical.RemoteControl.Build.ParseError do
+  alias Lexical.Document
+  alias Lexical.Plugin.V1.Diagnostic.Result
+
+  import Lexical.RemoteControl.Build.ErrorSupport,
+    only: [context_to_position: 1, position: 2, position: 4]
+
+  @elixir_source "Elixir"
+
+  # Parse errors happen during Code.string_to_quoted and are raised as SyntaxErrors, and TokenMissingErrors.
+  def to_diagnostics(
+        %Document{} = source,
+        context,
+        {_error, detail} = message_info,
+        token
+      )
+      when is_binary(detail) do
+    # NOTE: mainly for `unexpected token` errors `< 1.16`,
+    # its details consist of multiple lines, so it is a tuple.
+    detail_diagnostics = detail_diagnostics(source, detail)
+    error = message_info_to_binary(message_info, token)
+    error_diagnostics = to_diagnostics(source, context, error, token)
+    error_diagnostics ++ detail_diagnostics
+  end
+
+  def to_diagnostics(%Document{} = source, context, message_info, token)
+      when is_exception(message_info) do
+    to_diagnostics(source, context, Exception.message(message_info), token)
+  end
+
+  def to_diagnostics(%Document{} = source, context, message_info, token) do
+    end_line_fn =
+      if Features.details_in_context?() do
+        &build_end_line_diagnostics_from_context/4
+      else
+        &build_end_line_diagnostics/4
+      end
+
+    start_line_fn =
+      if Features.details_in_context?() do
+        &build_start_line_diagnostics_from_context/4
+      else
+        &build_start_line_diagnostics/4
+      end
+
+    parse_error_diagnostic_functions = [
+      end_line_fn,
+      start_line_fn,
+      &build_hint_diagnostics/4
+    ]
+
+    Enum.flat_map(parse_error_diagnostic_functions, & &1.(source, context, message_info, token))
+  end
+
+  defp build_end_line_diagnostics_from_context(
+         %Document{} = source,
+         context,
+         message_info,
+         token
+       ) do
+    message =
+      if String.starts_with?(message_info, "unexpected") do
+        ~s/#{message_info}#{token}, expected "#{context[:expected_delimiter]}"/
+      else
+        "#{message_info}#{token}"
+      end
+
+    if context[:end_line] do
```

Comment:
I feel like this should be happening in the location code

Reply:
And any naming suggestions?

Reply:
```elixir
case Location.extract_range(context) do 
  {:ok, %Range{end: end_pos}} -> [Result.new(source.uri, pos, end_pos, message, :error, @elixir_source)]
   _ -> []
 end
```